### ISSUE TITLE
planner: utilize the order property of prefix columns in primary key

### ISSUE THREAD

**Issue by qw4990** [pingcap/tidb#48241]
## Enhancement
For the case below, the order property of the primary key is not completely utilized. 
Actually, we can optimize the `FullScan` to a `RangeScan`, whose range should be `[the_first_value_of_a_*, the_second_value_of_a_*)`:
![677f2b6b-6a1f-481d-b457-e6060ec78a32](https://github.com/pingcap/tidb/assets/7499936/2913abb4-c876-46b7-881b-5efa8f7f2203)

Another similar case:

```
select distinct a, c from t order by a limit 1:
+------------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------------+
| id                           | estRows  | task      | access object | operator info                                                                                       |
+------------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------------+
| TopN_9                       | 1.00     | root      |               | test.t.a, offset:0, count:1                                                                         |
| └─HashAgg_18                 | 8000.00  | root      |               | group by:test.t.a, test.t.c, funcs:firstrow(test.t.a)->test.t.a, funcs:firstrow(test.t.c)->test.t.c |
|   └─TableReader_19           | 8000.00  | root      |               | data:HashAgg_14                                                                                     |
|     └─HashAgg_14             | 8000.00  | cop[tikv] |               | group by:test.t.a, test.t.c,                                                                        |
|       └─TableFullScan_17     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                      |
+------------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------------+
```

**Comment by winoros** [comment]
That we use a newly-implemented TopN to control the break time will be better than letting TableScan decide it.

**Comment by jiyfhust** [comment]
It is greate enhancement for some types of "order by" sql and very usefull.
It can be expanded to more sql typs other than above, and even to prefixIndex scenaries.
I think they have the common characteristics:
```
alter table t add index idx(col_a, col_b, col_c, col_d(32), col_e, ....);
// table t have a multi-column index named idx.
select * from t where [col_a condition, con_b condition, col_c condition] order by col_a, col_b, col_c limit m, n
```
1. the order by items have commom column parts with index items, and they have same order and are continuous in index postition, we here name the the same parts as common_parts.
2. the common_parts ends if:
    * all index items is included in common_parts
    * the order by item meets a diffrent item with index item
    * the order by item meets a item that is a prefix_index_part of index
3. the related columns in where conditions are all all in common_parts.

The case above, we can use the index to make a keep order scan to filter data we needed.

If 3 not satisfied, it is also helpfull if we can judge it cost less when we use indxeLoopup than full scan.

```
select * from t where col_b = "xxx" order by col_a, col_c limit m, n;

same as

select * from t where col_b = "xxx" order by col_a, col_b, col_c limit m, n;

they both have common_parts col_a, col_b, col_c.
```
When "the order by item meets a item that is a prefix_index_part of index", the common_parts should include the prefix_part column but with prefix data length.

There is a post in askTug related with this topic:
https://tidb.net/blog/524783fa.

**Comment by winoros** [comment]
> It is a great enhancement for some types of "order by" sql and very usefull. It can be expanded to more sql typs other than above, and even to prefixIndex scenaries. I think they have the common characteristics:
> 
> ```
> alter table t add index idx(col_a, col_b, col_c, col_d(32), col_e, ....);
> // table t have a multi-column index named idx.
> select * from t where [col_a condition, con_b condition, col_c condition] order by col_a, col_b, col_c limit m, n
> ```
> 
> 1. the order by items have commom column parts with index items, and they have same order and are continuous in index postition, we here name the the same parts as common_parts.
> 2. the common_parts ends if:
>    
>    * all index items is included in common_parts
>    * the order by item meets a diffrent item with index item
>    * the order by item meets a item that is a prefix_index_part of index
> 3. the related columns in where conditions are all all in common_parts.
> 
> The case above, we can use the index to make a keep order scan to filter data we needed.
> 
> If 3 not satisfied, it is also helpfull if we can judge it cost less when we use indxeLoopup than full scan.
> 
> ```
> select * from t where col_b = "xxx" order by col_a, col_c limit m, n;
> 
> same as
> 
> select * from t where col_b = "xxx" order by col_a, col_b, col_c limit m, n;
> 
> they both have common_parts col_a, col_b, col_c.
> ```
> 
> When "the order by item meets a item that is a prefix_index_part of index", the common_parts should include the prefix_part column but with prefix data length.
> 
> There is a post in askTug related with this topic: [tidb.net/blog/524783fa](https://tidb.net/blog/524783fa).

Yes, implementing a new TopN or integrating it with the current TopN to support the prefix partial order would be a good idea.
